README.md
[![ci-cd](https://github.com/UBC-MDS/easyphysics/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/easyphysics/actions/workflows/ci-cd.yml) [![codecov](https://codecov.io/gh/UBC-MDS/easyphysics/branch/main/graph/badge.svg?token=7jGlm8VZMi)](https://codecov.io/gh/UBC-MDS/easyphysics) [![Documentation Status](https://readthedocs.org/projects/easyphysics/badge/?version=latest)](https://easyphysics.readthedocs.io/en/latest/?badge=latest)


# easyphysics

A package with useful physics formulas to make physics easy and fun for users! It uses four functions to easily calculate four classic physics theories. By making the functions concise and supporting graphs to demonstrate the theories, physics beginners might find learning physics fun and not daunting! All functions in this package require only arguments in numeric format; no dataset files are needed. `easyphysics` can also generate easy-to-understand visualizations to further reveal the physics effects.  

## Installation

```bash
$ pip install easyphysics
```

## Usage

The package has four functions which can provide solutions for Physics Equations. Input will be taken from the User and the solutions will be provided by the functions. The four functions are as follows: 


- `freefall()` calculates the time it takes for a falling object using the equation of motion height = 1/2*gt^2, given the height and gravity of the free fall. It returns the time it takes for the free fall, and a plot compares the time of the free fall on different planets with a list. The distance traveled by the falling object (height) and the acceleration of gravity (g, default = 9.8) are the function’s arguments. 
- `gravitational_energy()` calculates the energy possessed or acquired by an object due to a change in its position when it is present in a gravitational field  = m*g*h
- `kinetic_energy()` calculates the Kinetic Energy of an object. When work is done on an object, energy is transferred, and the object moves with a new constant speed. We call the energy that is transferred kinetic energy, and it depends on the mass and speed achieved. The kinetic energy equation is given as: KE = 1/2*m*v^2, Where KE is the kinetic energy, m is the body’s mass, and v is the body’s velocity.
- `static_friction_ground()` calculates the friction force for static object. The formula is fr = mu * N, where 
    the mu is the coefficient of friction which incorporating the characteristics of the surface.

```python
from easyphysics.freefall import freefall
# time, plot = freefall(height, g = 9.8)
time, plot = freefall(10, g = 15)

```

```python
from easyphysics.kinetic_energy import kinetic_energy
# KE = kinetic_energy(mass, velocity)
KE = kinetic_energy(0.6, 3)

```

```python
from easyphysics.gravitational_energy import gravitational_energy
# gravitational_energy(m,h, g = 9.80665)
ge = gravitational_energy(10, 1000, g = 9.8)

```

```python
from easyphysics.static_friction import static_friction_ground
from easyphysics.static_friction import plot_frict
# static_friction_ground(mu, m, g = 9.8)
force = static_friction_ground(0.2, 2)
plot_frict(0.2,2,9.8)

```
## Fitting into the Python ecosystem

A similar well-known package [sympy](https://github.com/sympy/sympy) is available, which has a [module in physics](https://docs.sympy.org/latest/reference/public/physics/index.html).`sympy` is a library for symbolic computation. The physics module focuses on solving advanced physics problems such as Hydrogen Wavefunction and Quantum physics.

`easyphysics` targets physics beginners rather than professionals, such as middle schoolers and people new to physics. It uses four functions to easily calculate four classic physics theories. By making the functions concise and supporting graphs to demonstrate the theories, physics beginners might find learning physics fun and not daunting! All functions in this package require only arguments in numeric format; no dataset files are needed.  `easyphysics` can also generate easy-to-understand visualizations to further reveal the physics effects.  The specific functions and visual components make  `easyphysics` different from other packages.

## Contributing

`easyphysics` was created by Revathy Ponnambalam, Nikita Susan Easow, Yaou Hu and Mengjun Chen.
Interested in contributing? Check out the [contributing guidelines](CONTRIBUTING.md). Please note that this project is released with a [Code of Conduct](CONDUCT.md). By contributing to this project, you agree to abide by its terms.

## License

`easyphysics` was created by Revathy Ponnambalam, Nikita Susan Easow, Yaou Hu and Mengjun Chen. It is licensed under the terms of the MIT license.

## Credits

`easyphysics` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.2.11 (2023-02-08)


## v0.2.10 (2023-02-05)


## v0.2.9 (2023-02-04)


## v0.2.8 (2023-02-04)


## v0.2.7 (2023-02-04)


## v0.2.6 (2023-02-02)


## v0.2.5 (2023-01-31)


## v0.2.4 (2023-01-30)


## v0.2.3 (2023-01-29)


## v0.2.2 (2023-01-29)


## v0.2.1 (2023-01-29)


## v0.2.0 (2023-01-29)
### Feature
* Adding comment! ([`5be5f8b`](https://github.com/UBC-MDS/easyphysics/commit/5be5f8ba9c45c6141f0dd3bf7c719a69ba278896))

### Fix
* Adding comment again! ([`08a6cd6`](https://github.com/UBC-MDS/easyphysics/commit/08a6cd6b10210cfb4dc6c999b610b8dc9ba14e15))
* Removing comment! ([`b5d812d`](https://github.com/UBC-MDS/easyphysics/commit/b5d812d8e03e15d2e5744c2a18159e5eb1e82c49))

## v0.1.1 (2023-01-29)


## v0.1.0 (12/01/2023)

- First release of `easyphysics`!

tests/test_easyphysics.py
import sys
#sys.path.append("..\\src\\")
print(sys.path)
from easyphysics.kinetic_energy import kinetic_energy
from easyphysics.gravitational_energy import gravitational_energy
from easyphysics.freefall import freefall
from easyphysics.static_friction import static_friction_ground
import matplotlib
import pytest

def test_kinetic_energy():
    """Test the kinetic energy function"""
    # input values to function - m, v
    m = 10
    v = 2.3
    # expected output value
    expected = 26.45
    # enter this value in kinetic_energy function
    actual = kinetic_energy(m, v)
    # if answer is not expected, test fails and returns appropriate error message
    assert actual == expected, "Kinetic Energy is wrong! Check again!"
    
def test_freefall():
    """ Test freefall function time and plot.
    if the function returns the expected data type,
    and if the function returns the expected value.
    """
    
    # test three different input combinations
    # test if the function returns the calculation as expected
    # test if the function works with different inputs
    expected_time1 = 1.4286
    actual_time1, plot1 = freefall(10)
    expected_time2 = 4.47
    actual_time2, plot2 = freefall(100, g = 10)
    expected_time3 = 0.3162
    actual_time3, plot3 = freefall(5, g = 100)    
    assert round(expected_time1,2) == round(actual_time1,2), "freefall calculation not correct"
    assert round(expected_time2,2) == round(actual_time2,2), "freefall calculation not correct"
    assert round(expected_time3,2) == round(actual_time3,2), "freefall calculation not correct"
    
    # test the output has a matplotlib plot
    # test the output for different inputs
    assert type(plot1) == matplotlib.figure.Figure, \
           "Not a plot returned!"
    assert type(plot2) == matplotlib.figure.Figure, \
           "Not a plot returned!"
    assert type(plot3) == matplotlib.figure.Figure, \
           "Not a plot returned!"
    
    # test inputting height as not a number
    # TypeError raises if height is not input as a number
    with pytest.raises(TypeError):
        freefall('a')
    
    # test input g as not a number
    # TypeError raises if g is not input as a number
    with pytest.raises(TypeError):
        freefall(100, g = 'g')
        
def test_gravitation_energy():
    """ Test Gravitational Energy"""
    # entering expected output values for gravitation_energy function
    expected1 = 245.00
    expected2 = 2157.463 
    # performing operation with set values
    actual1 = gravitational_energy(2.5,10,g = 9.8)
    actual2 = gravitational_energy(10,22,g = 9.80665)
    # comparing if the computed value matches the actual value - if not, throws a relevant error
    assert round(actual1,2) == round(expected1,2), "Gravitation Energy calculated is not right! Do Check inputs"
    assert round(actual2,2) == round(expected2,2), "Gravitation Energy calculated is not right! Do Check inputs"
    

    
def test_friction():
    """ Test Friction Force on Ground,
    and if the function returns the expected value.
    """
    expected1 = 3.92
    expected2 = 6.6
    actual1 = static_friction_ground(mu=0.2, m=2, g=9.8)
    actual2 = static_friction_ground(mu=0.3, g = 10, m=2.2)
    
    assert round(actual1,2) == round(expected1,2), "Friction force calculated is not right! Do Check inputs"
    assert round(actual2,2) == round(expected2,2), "Friction force calculated is not right! Do Check inputs"


src/easyphysics/freefall.py
import numpy as np
import matplotlib.pylab as plt


def freefall(height, g=9.8):
    """ calculates the time it takes for a falling object
    using the equation of motion height = 1/2*gt^2, 
    given the height and gravity of the free fall. 
    It returns the time it takes for the free fall, 
    and a plot compares the time of the free fall 
    on different planets with a list. 
    The distance traveled by the falling object (height)
    and the acceleration of gravity (g, default = 9.8) 
    are the function’s arguments.
    
    Parameters
    ----------
    height : numeric
        A numeric value that is the distance of the free fall in meters.
    g: numeric, optional
       The gravity of the free fall (m/s^2), default value = 9.8.
    Returns
    -------
    time: float
        A float that is the time it takes for the free fall in seconds.
    plot: matplotlib.figure.Figure,
        A matplotlib.figure.Figure that compares time of a free fall with
        the height indicated by user on different planets.
    Examples
    --------
    >>> freefall(10, g=9.8)
    1.4286
    >>> freefall(10, g=274)
    0.2702
    """
    if (type(height) == int or type(height) == float) & (type(g) == int or type(g) == float):
        time = np.sqrt(2*height/g)
        gdict = {'Moon': 0.17,
                 'Mars': 0.38,
                'Uranus': 0.89,
                'Venus': 0.90,
                'Saturn': 1.07,
                'Sun': 27.9}
        gdict['Your planet'] = g/9.8
        sorted_gdict = dict(sorted(gdict.items(), key=lambda x:x[1]))
        planet = []
        gravity = []
        times = []
        for value in sorted_gdict.values():
            gravity.append(value)
        for key in sorted_gdict.keys():
            planet.append(key)
        for value in sorted_gdict.values():
            times.append(np.sqrt(2*height/(9.8*value)))
        plt.scatter(planet,times)
        plt.ylabel("Time for the free fall in seconds")
        plt.xlabel("Planets")
        plt.title("Time for the free fall on different planets for the height you entered")
        plt.figure(figsize = (10,10))
        plot = plt.figure()
    else:
        raise TypeError('height and g has to be numeric!')
    return time, plot


src/easyphysics/static_friction.py
import altair as alt
import numpy as np
import pandas as pd

def static_friction_ground(mu, m, g = 9.8):
    """ calculating the friction force for static object. The formula is fr = mu * N, where 
    the mu is the coefficient of friction which incorporating the characteristics of the surface.
    For example, if you want to calculate the friction when you put your phone on a table, then the coefficient of friction is 
    depending on the matirial of the table. N is the normal force which means the perpendicular force of the surface which object 
    touching on. In our case, it would be m*g since it is on the ground. fr = mu * m * g. 
    
    Parameters
    ---------------
    mu : numeric 
        coefficient of friction
    m: numeric
        mass (kg)
    g: numeric, optimal     
        gravity on Earth
    
    return
    ---------------
    force: numeric
        static friction force
        
    Examples
    ---------------
    >>> static_friction_ground(0.2, 2)
    3.92
    >>> static_traction_ground(0.2, 2, g = 10)
    4
    """
    #checking the input values to be the right 
    if (type(mu) == int or type(mu) == float) and (type(g) == int or type(g) == float) and (type(m) == int or type(m) == float):
        result = mu * m * g
    else:
        raise TypeError('height and g has to be numeric!')

    return result

def plot_frict(m, mu, g):
    """
    Ploting the trend line for firction force in different masses 
    
    parameter
    ------------
    mu : numeric 
        coefficient of friction
    m: numeric
        mass (kg)
    g: numeric, optimal     
        gravity on Earth
    
    return
    ---------------
    plot
        
    Examples
    ---------------
    >>> static_friction_ground(0.2, 2)
    line plot 
    >>> static_traction_ground(0.2, 2, g = 10)
    line plot
    """
    
    if (type(mu) == int or type(mu) == float) and (type(g) == int or type(g) == float) and (type(m) == int or type(m) == float):
        
        masses = np.arange(0, m, 0.1)

        friction_line = [static_friction_ground(mu = mu, m=i.item() , g=g) for i in masses]

        source = pd.DataFrame({
            'mass': masses,
            'friction': friction_line
            })
        line = alt.Chart(source).mark_line().encode(
            x='mass',
            y='friction',
            tooltip = ['mass', 'friction']
        ).properties(
        title = "The static friction as mass increase"
        )
    else:
        raise TypeError('mu, mass and g has to be numeric!')
        
    return line
